### DISCUSSION TITLE
WiCAN how to receive SOC from Smart EQ forfour

### DISCUSSION THREAD

**Discussion by martin8310** [meatpiHQ/wican-fw#115] [Q&A]
Hello everyone,

Could someone explain me how can I receive this data over MQTT, please.

The following information is from the Wican Monitor website:

Byte 3: This is the third byte of the data. Represented 14: This is the value 14 represented in hexadecimal. So I suspect that 14 hex (20 percent) is my SOC from my Smart.

````
ID		Type	Len	Data						Time	Count
654h	        Std	8	00 00 00 14 FF C1 92 80		                466	390
````

How can I query this information in the bottom fields:
![image](https://github.com/meatpiHQ/wican-fw/assets/62466200/1daf3e34-3579-48f0-953a-7182a3044325)

How should the query look according to your schema?:
````
{"bus":0,"type":"tx","frame":[{"id":123,"dlc":8,"rtr":false,"extd":true,"data":[1,2,3,4,5,6,7,8]},{"id":124,"dlc":8,"rtr":false,"extd":true,"data":[1,2,3,4,5,6,7,8]}]}
````

Thank you in advence.

Regards
Martin

**Comment by meatpiHQ** [comment]
@martin8310 

> Byte 3: This is the third byte of the data. Represented 14: This is the value 14 represented in hexadecimal. So I suspect that 14 hex (20 percent) is my SOC from my Smart.

Does the bus sends this frame continuously? if you can see it in the CAN monitor without sending than it must be the case.

**Reply by martin8310** [reply]
Yes, that will be the case. But I don't receive anything via MQTT. Only the online/offline status.

**Reply by meatpiHQ** [reply]
in this case you can add a filter for that frame.

ID: 1620
PID: -1
Start Bit: 0
Bit Length:1
Expression: B4
cycle: 1000

This will send the value of B4 to MQTT. Make sure you reboot the device after you use the CAN monitor.

**Reply by martin8310** [reply]
Thanks, I'll give it a try. When I set the setting MQTT elm327 log: to Disabled, the live data also arrived via MQTT, but unfortunately too much ;-) How can I limit the data stream?

**Reply by meatpiHQ** [reply]
Adding a filter will block all frames except the ones you added in the filter, and the number of frames you will receive of the filtered ID will depend on the value you entered in Cycle, 1000 is 1sec.

**Reply by martin8310** [reply]
So, I set the values, but a message comes from WiCAN "PID must be a number between 0 and 7".
What about the index field, should I leave it empty?

´´´
ID: 1620
PID: -1
Start Bit: 0
Bit Length:1
Expression: B4
cycle: 1000
´´´

**Reply by martin8310** [reply]
"PID -1 works after all. But I had to set the Index field to 0! No idea if that's correct.

**Reply by meatpiHQ** [reply]
If you set the PID to -1 the index value is ignored. So it should work.

**Reply by martin8310** [reply]
I am using firmware version v2.98. If I leave the index field empty, I cannot add an entry using the "ADD ID" button.

**Reply by meatpiHQ** [reply]
You can put any value it doesn't matter.

**Reply by martin8310** [reply]
Does this mean I need to set "MQTT elm327 log:" to Enabled for the filter in the Add ID line to work, or does MQTT elm327 log: need to be disabled?

**Reply by meatpiHQ** [reply]
No the ELM327 log has to be disabled

**Reply by martin8310** [reply]
![image](https://github.com/meatpiHQ/wican-fw/assets/62466200/bc3ac4aa-343c-4d8c-a250-e655f0506631)

when it is saved (Submit Changes), it looks like this:

![image](https://github.com/meatpiHQ/wican-fw/assets/62466200/cc773dd9-6c33-41f4-8933-2b679e130457)

I think my meatPi needs a complete reset ;-)

**Reply by meatpiHQ** [reply]
Ah! Did you click store button? It's above the "Add id" button

**Reply by martin8310** [reply]
no

**Reply by martin8310** [reply]
So my dear, now it works. Expression had to be on B3, so the battery level is also output immediately as a decimal number in MQTT. Thank you ;-)

{"SOC2": 97.000000}

![image](https://github.com/meatpiHQ/wican-fw/assets/62466200/0c00ac5a-e0a8-4e4e-9e0b-d32ca20e24ca)

**Reply by meatpiHQ** [reply]
Yes that's right B3, my bad.

**Comment by martin8310** [comment]
Hello its me again,

Can someone tell me how to extract the ODO value from this row? The formula is usually like this: 

`SetValue((float) (CAN_UINT32(2)>>4) / 100);`

This is the address at which the value resides:
```
ID	Type	Len		Data				Time	Count
5D7h	Std	8		00 00 02 4B 1C 20 FA 00		108	189
```

It schould be 24048km.

Thanky in advance.

regards Martin

**Reply by martin8310** [reply]
The Solution for that is: `(B2*16.777.216+B3*65536+B4*256+B5)/16/100`
the shorter spelling: `[B2:B5]/16/100`

**Comment by martin8310** [comment]
Would be great to get the Smart EQ added to the WiCAN library of cars. :-)

**Reply by meatpiHQ** [reply]
@martin8310 Yes! I would love to add it, if you can help with some logs from car scanner?

**Reply by martin8310** [reply]
The car is not supported by Car Scanner, so it will be difficult.

But what I have is this:
![image](https://github.com/user-attachments/assets/49b2b20e-f3c5-4890-bc2d-6928e826c15a)

**Reply by martin8310** [reply]
1620 is sent continuously. 1495 only when the engine is on.

**Reply by martin8310** [reply]
And we have that from OVMS :
```
0x392:
  StandardMetrics.ms_v_env_hvac->SetValue((CAN_BYTE(1) & 0x40) > 0);

0x42E: // HV Voltage
  StandardMetrics.ms_v_bat_voltage->SetValue((float) ((CAN_UINT(3)>>5)&0x3ff) / 2); // HV Voltage
  StandardMetrics.ms_v_bat_temp->SetValue(((c >> 13) & 0x7Fu) - 40); // HVBatteryTemp
  StandardMetrics.ms_v_charge_climit->SetValue((c >> 20) & 0x3Fu); // MaxChargingNegotiatedCurrent

0x5D7: // Speed, ODO
  StandardMetrics.ms_v_pos_speed->SetValue((float) CAN_UINT(0) / 100);
  StandardMetrics.ms_v_pos_odometer->SetValue((float) (CAN_UINT32(2)>>4) / 100);

0x5de:
  StandardMetrics.ms_v_env_headlights->SetValue((CAN_BYTE(0) & 0x04) > 0);
  StandardMetrics.ms_v_door_fl->SetValue((CAN_BYTE(1) & 0x08) > 0);
  StandardMetrics.ms_v_door_fr->SetValue((CAN_BYTE(1) & 0x02) > 0);
  StandardMetrics.ms_v_door_rl->SetValue((CAN_BYTE(2) & 0x40) > 0);
  StandardMetrics.ms_v_door_rr->SetValue((CAN_BYTE(2) & 0x10) > 0);
  StandardMetrics.ms_v_door_trunk->SetValue((CAN_BYTE(7) & 0x10) > 0);

0x654: // SOC(b)
  StandardMetrics.ms_v_bat_soc->SetValue(CAN_BYTE(3));
  StandardMetrics.ms_v_door_chargeport->SetValue((CAN_BYTE(0) & 0x20)); // ChargingPlugConnected
  StandardMetrics.ms_v_charge_duration_full->SetValue((((c >> 22) & 0x3ffu) < 0x3ff) ? (c >> 22) & 0x3ffu : 0);
  StandardMetrics.ms_v_bat_range_est->SetValue((c >> 12) & 0x3FFu); // VehicleAutonomy

0x65C: // ExternalTemp
  StandardMetrics.ms_v_env_temp->SetValue((CAN_BYTE(0) >> 1) - 40); // ExternalTemp ?

0x658: //
  StandardMetrics.ms_v_bat_soh->SetValue(CAN_BYTE(4) & 0x7Fu); // SOH ?
  isCharging = (CAN_BYTE(5) & 0x20); // ChargeInProgress

0x673:
  if (CAN_BYTE(2) != 0xff)
    StandardMetrics.ms_v_tpms_pressure->SetElemValue(MS_V_TPMS_IDX_RR, (float) CAN_BYTE(2)*3.1);
  if (CAN_BYTE(3) != 0xff)
    StandardMetrics.ms_v_tpms_pressure->SetElemValue(MS_V_TPMS_IDX_RL, (float) CAN_BYTE(3)*3.1);
  if (CAN_BYTE(4) != 0xff)
    StandardMetrics.ms_v_tpms_pressure->SetElemValue(MS_V_TPMS_IDX_FR, (float) CAN_BYTE(4)*3.1);
  if (CAN_BYTE(5) != 0xff)
    StandardMetrics.ms_v_tpms_pressure->SetElemValue(MS_V_TPMS_IDX_FL, (float) CAN_BYTE(5)*3.1);
```

**Reply by martin8310** [reply]
Is it helpful, or do you still need car scanner information?

The current smart uses ISO 14230-4 to send ECU data to the OBD, most European and Asian manufactured cars use ISO 9141-2.

**Reply by meatpiHQ** [reply]
@martin8310 look like your car is broadcasting on the bus, this is not yet supported in the vehicle profile. I will try add support for this in the coming weeks.

**Comment by snorre-k** [comment]
@meatpiHQ Did you make any progress implementing the feature of automating broadcasted CAN messages in vehicle profiles? I also would really like this feature.

**Reply by meatpiHQ** [reply]
No not yet. But you can use the CAN filters for now.

https://meatpihq.github.io/wican-fw/mqtt/usage

**Comment by spuky** [comment]
@meatpiHQ does the profile for **Smart EQ** also work for the **Smart EQ forfour** since the discussion here uses totally different pids and elm327 as a protocol ? Just asking to not spend to much time trying to use...the profile ?

**Reply by meatpiHQ** [reply]
@spuky it's hard to say if it will work, I need to look at vehicle logs
